File: 1298-reverse-substrings-between-each-pair-of-parentheses/reverse-substrings-between-each-pair-of-parentheses.go
func reverseParentheses(s string) string {
    var stack []rune
    for _, ch := range s {
        if ch == ')' {
            var temp []rune
            for len(stack) > 0 && stack[len(stack)-1] != '(' {
                temp = append(temp, stack[len(stack)-1])
                stack = stack[:len(stack)-1]
            }
            if len(stack) > 0 {
                stack = stack[:len(stack)-1]
            }
            stack = append(stack, temp...)
        } else {
            stack = append(stack, ch)
        }
    }
    return string(stack)
}

File: 1298-reverse-substrings-between-each-pair-of-parentheses/reverse-substrings-between-each-pair-of-parentheses.cpp

class Solution {
public:
    string reverseParentheses(string s) {
        stack<char> n;
        for (char c : s) {
            if (c == ')') {
                string p;
                while (n.top() != '(') {
                    p += n.top();
                    n.pop();
                }
                n.pop();
                for (char ch : p) {
                    n.push(ch);
                }
            } else {
                n.push(c);
            }
        }
        
        string a;
        while (!n.empty()) {
            a += n.top();
            n.pop();
        }
        
        reverse(a.begin(), a.end());
        return a;
    }
};
